### ISSUE TITLE
Help us clean up Bash-it!

### ISSUE THREAD

**Issue by NoahGorny** [Bash-it/bash-it#1696]
Hello folks!
We have now integrated `pre-commit` hooks in order to improve the codebase quality. and now we need to start going through the files. We invite anyone who wants to contribute to this effort to try this out:
* Choose a file in the project, which is not present in `clean_files.txt`
* Add the path from the root of the project to the file in `clean_files.txt`
* Run ``pre-commit run --files `cat clean_files.txt` ``, and fix the issues
* When `pre-commit` runs successfully, feel free to open a PR
* Hooray, you helped us make Bash-it better! :smile:

**Comment by tbhaxor** [comment]
Pre commit can be ignore using `--no-verify` and it will also frustrate the contributor when they are working on too many commits. So I would recommend you to add these checking tasks in the github workflow

**Comment by tbhaxor** [comment]
I will help you refactoring bash-it

**Comment by NoahGorny** [comment]
> Pre commit can be ignore using `--no-verify` and it will also frustrate the contributor when they are working on too many commits. So I would recommend you to add these checking tasks in the github workflow

It is indeed pretty frustrating, but checked in the CI (see `lint` stage)

Also see #1746 for better instructions

Thanks you @tbhaxor for helping us :heart:

**Comment by tbhaxor** [comment]
Hey @NoahGorny I found so much inconsistent code in plugins directory. :smile:

**Comment by NoahGorny** [comment]
> Hey @NoahGorny I found so much inconsistent code in plugins directory. smile

Hi @tbhaxor, there is indeed a lot of inconsistent code in the plugins (and in the rest of the codebase as well). It does not mean we should give up though! We can continue improving the codebase, one PR at a time :smile:

**Comment by cornfeedhobo** [comment]
@tbhaxor haha indeed. Luckily, @NoahGorny has been putting in an amazing amount of effort into figuring out our contribution guidelines and tooling. I see a bright future ahead :)

**Comment by tbhaxor** [comment]
@NoahGorny Is anything left in this?

**Comment by NoahGorny** [comment]
> @NoahGorny Is anything left in this?

There are actually many things that are still not cleaned up- most of the plugins, the `lib` directory, and also the install script.
If you want on those things, it would be really great :smile:

**Comment by cevhyruz** [comment]
@tbhaxor hi, I was wondering if you have already started on this? I'd also like to lend a hand if you're not already done

**Comment by tbhaxor** [comment]
> > @NoahGorny Is anything left in this?
> 
> There are actually many things that are still not cleaned up- most of the plugins, the `lib` directory, and also the install script.
> If you want on those things, it would be really great

@NoahGorny Actually I am down with covid so I won't be able to work on that. Could you guys take care of it?

**Comment by cevhyruz** [comment]
Sorry to hear that, hope you get well soon.

I'll be starting to work on it then.

@NoahGorny, @tbhaxor
Just a quick question, should I need a separate PR for every file / directory that has been cleaned
or should I just commit everything in a single PR?

**Comment by NoahGorny** [comment]
Well @tbhaxor, I hope you will get through this, you have my support!

> Sorry to hear that, hope you get well soon.
> 
> I'll be starting to work on it then.
> 
> @NoahGorny, @tbhaxor
> Just a quick question, should I need a separate PR for every file / directory that has been cleaned
> or should I just commit everything in a single PR?

I prefer that you split your work into separate PRs, this way is easier to check and merge.
If you have any troubles feel free to contact me.

Thanks for your decision to help Bash-it! You rock :smile:

**Comment by seefood** [comment]
seems like I can't compile the pre-commit hooks without a bit of help (see bottom)
```
        as: unrecognized option '--gdwarf-5'
        error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
        ----------------------------------------
    ERROR: Command errored out with exit status 1: /home/ira/.cache/pre-commit/repop169ejxs/py_env-python3.9/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-707uv1_t/python-levenshtein_1a27053a5a29481fbbdcee325f499a89/setup.py'"'"'; __file__='"'"'/tmp/pip-install-707uv1_t/python-levenshtein_1a27053a5a29481fbbdcee325f499a89/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-oab8hg3w/install-record.txt --single-version-externally-managed --compile --install-headers /home/ira/.cache/pre-commit/repop169ejxs/py_env-python3.9/include/site/python3.9/python-Levenshtein Check the logs for full command output.
    
Check the log at /home/ira/.cache/pre-commit/pre-commit.log
```
### version information

```
pre-commit version: 2.17.0
git --version: git version 2.32.0
sys.version:
    3.9.7 (default, Sep 10 2021, 14:59:43) 
    [GCC 11.2.0]
sys.executable: /usr/bin/python
os.name: posix
sys.platform: linux
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=21.10
```

setting `export CC=gcc-8` does the trick (a search says any version gcc-10 and earlier would do)

**Comment by akinomyoga** [comment]
There are many fixes suggested in discussing the refactoring (and other fixes) and marked as "should be fixed in separate PRs", but left unfixed and forgotten. This is a list of such items that I remember now:

- [ ] Remove legacy loading: https://github.com/Bash-it/bash-it/pull/2097#issuecomment-2473310024
- [ ] Utility to split lines: https://github.com/Bash-it/bash-it/pull/2282#discussion_r1930235467
- [x] plugin/man: Localize env variables: https://github.com/Bash-it/bash-it/pull/2289#pullrequestreview-2626177126
- [x] completion/gradle: Fix a `find` expression to enclose `expr -o expr` in `\( ... \)`: https://github.com/Bash-it/bash-it/pull/2294#discussion_r2009045990
- [x] https://github.com/Bash-it/bash-it/pull/2298#discussion_r2068572837
- [x] https://github.com/Bash-it/bash-it/pull/2048#issuecomment-2841770395
- [x] https://github.com/Bash-it/bash-it/pull/2060#discussion_r2068564223
- [x] https://github.com/Bash-it/bash-it/pull/2302#discussion_r2068557809
- [ ] Rework migration to `lib/history.bash` https://github.com/Bash-it/bash-it/pull/1951#discussion_r2079623415
- [ ] `case :$HISTCONTROL: in` https://github.com/Bash-it/bash-it/pull/1951#issuecomment-2862862442
- [ ] Documentation of `auto*` in `HISTCONTROL` https://github.com/Bash-it/bash-it/pull/1951#pullrequestreview-2824836982
- [ ] Prompt escaping https://github.com/Bash-it/bash-it/pull/2315#discussion_r2126491153
- [ ] CHA, CUD, or nothing https://github.com/Bash-it/bash-it/pull/2326#discussion_r2198908820
- [ ] `COMPREPLY=("${newline_separated_scalar[@]}")`: https://github.com/Bash-it/bash-it/pull/2063#pullrequestreview-3149262316
- [ ] Bash version check in `pack.plugin.bash` is unnecessary https://github.com/Bash-it/bash-it/pull/2358#discussion_r2429017401

**Comment by seefood** [comment]
I see merged PRs on all the items are they all indeed closed issues? is this a good time to mint a release?

**Comment by akinomyoga** [comment]
> I see merged PRs on all the items are they all indeed closed issues?

Of course, they are closed issues by definition. The above reply lists the changes *skipped* in the PRs and issues.

> is this a good time to mint a release?

These are still remaining problems, though you can make a release before fixing these.

> * [x]  [theme/liquidprompt: lint and use `$THEME_CLOCK_FORMAT` #2048 (comment)](https://github.com/Bash-it/bash-it/pull/2048#issuecomment-2841770395)

Was the license issue resolved?

> * [x]  Prompt escaping [Clean themes A-L #2315 (comment)](https://github.com/Bash-it/bash-it/pull/2315#discussion_r2126491153)

I don't think the mentioned problem was resolved.

> * [x]  CHA, CUD, or nothing [Fix for 2323 issue #2326 (comment)](https://github.com/Bash-it/bash-it/pull/2326#discussion_r2198908820)

This is also not addressed.

**Comment by seefood** [comment]
ohhh... my bad. I should have read extra carefully.
as for license - I say it's not worth bothering that much with one of dozens of themes. I'll have a look and deside fiwe keep it at all.

**Comment by akinomyoga** [comment]
> as for license - I say it's not worth bothering that much with one of dozens of themes.

We are infringing the copyright of Liquidprompt. We need to put effort into resolving the copyright infringement, or we are criminals.

**Comment by akinomyoga** [comment]
To resolve it, we can mention the license of `theme/liquidprompt` in the README, and include the license file in `themes/liquidprompt/`.

**Comment by seefood** [comment]
true, but maybe the best way is to DL the files locally by the user's own machine, IFF they decide they want it.

**Comment by seefood** [comment]
finally got around to look at it, seems to me like we are expecting liquidprompt to be in /vendor and then wrap around it, but it's not even in vendor. what am I missing here? seems like none of their code is really in our repo. did I miss something?

**Comment by seefood** [comment]
as for the man colour issue, I hope #2338 is a good fix.

**Comment by akinomyoga** [comment]
> finally got around to look at it, seems to me like we are expecting liquidprompt to be in /vendor and then wrap around it, but it's not even in vendor. what am I missing here? seems like none of their code is really in our repo. did I miss something?

Thank you for checking. OK, I checked the contents of the `liquidprompt` theme and realized that I misunderstood the original comment from `@gaelicWizard`. I'm sorry for making you do extra work.

**Comment by seefood** [comment]
> I'm sorry for making you do extra work.

nope. absolutely not. I need to know bash-it better. the fact I got merge rights here because the old guard has no time left for it means I can do merges and that's it. I am don't have your deep understanding of bash and I am not sure you don't have a better understanding than me of bash-it itself. that's why I am so grateful I have someone else to ask when I am not sure.

**Comment by akinomyoga** [comment]
Thank you for your kind words!

**Comment by akinomyoga** [comment]
This issue was automatically closed because #2373 included the string `Fixes #1696`, but #2373 doesn't fix all issues here. Could you please reopen this issue?

**Comment by seefood** [comment]
yeah, it was an automatic GH thing.
